### PR TITLE
#1214 - Warning on Discontinued reviews (Unassigned)

### LINKED ISSUE #1214: Show message to user unassigned to review on the ReviewPage
IMPORTANT: PSM wrote: Approved and recommended to be in initial MVP (02/06/2022)

### Reference 

**Final Requirements Matrix 2** - Customisations - Row 12 

### Requirement discussion
- [x] Fix https://github.com/openmsupply/conforma-web-app/issues/1216 (Not requested - but should helps to remove confusion)
It's an edge case, since usually the re-assignment is done because the original reviewer is not working on the review, but it happened during a workshop. and it's confusing to the person reviewing when try to submit and nothing happens.

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #1214 
Back-end PR: [837](https://github.com/openmsupply/conforma-server/pull/837)

Tests: Please use snapshot `dev/snapshots/Unassignment_warning_test` for testing this. It won't work currently in out angola snapshot since the actions on event ON_REVIEW_ASSIGN and ON_REVIEW_UNASSIGN are not correctly set/missing.

Once you load the `Unassignment_warning_test` snapshot please run `yarn migrate` to get changes on the back-end working.

Then test: 

1. load as `jess` and see the Review of http://localhost:3000/application/S-OSW-0002/review?tab=assignment
2. Open the review page, make changes if desired - but don't submit
3. Open an incognito browser and log in as `nicole` then in http://localhost:3000/application/S-OSW-0002/review?tab=assignment do Un-assignment of jess. 
4. In the other tab looking at the Review page, if you try to add a new review_element, click "Approve all", submit the review or reload it should show the Warning:
Screen Shot 2022-09-06 at 1 12 50 PM

No changes can be done - but you see the previous added elements:
Screen Shot 2022-09-06 at 1 13 08 PM

If this the other user (`nicole`) now Re-assigning back to user `jess`. Then the review should be able to load fine and continued.

### Changes
- Back-end: Fix activity_log functions that caused a problem when running action `ChangeStatus` to set the Review to DISCONTINUED on Un-assignment.
- Updated elements on page to not-editable when the Review status is not DRAFT. Strange that this was allowed before!

### CODE REVIEW

File: src/containers/Review/ReviewPage.tsx [hunk 1]
```diff
@@ -48,9 +48,11 @@ const ReviewPage: React.FC<{
   fullApplicationStructure: FullStructure
 }> = ({ reviewAssignment, previousAssignment, fullApplicationStructure }) => {
   const { strings } = useLanguageProvider()
+  const { serial } = fullApplicationStructure.info
   const {
     userState: { currentUser },
   } = useUserState()
+  const [reviewStatus, setReviewStatus] = useState(ReviewStatus.Submitted)
```
Comment (CarlosNZ) [inline src/containers/Review/ReviewPage.tsx:55]: Not entirely sure why this needs to be a state variable. Can't it just be derived from `reviewStructure` at any time?

I think it works to just declare it after the `useGetReviewStructureForSections` hook:

```ts
const { reviewStructure, error } = useGetReviewStructureForSections({
    reviewAssignment,
    reviewStructure: fullApplicationStructure,
  })

const reviewStatus = reviewStructure?.thisReview?.current.reviewStatus ?? ReviewStatus.Submitted
```

It'll still get updated whenever the value in `reviewStatus` changes (as reviewStatus will be updated from the hook).

File: src/containers/Review/ReviewPage.tsx [hunk 2]
```diff
@@ -68,19 +70,30 @@ const ReviewPage: React.FC<{
...
+    showCancel: false,
   })
 
   useEffect(() => {
-    if (!reviewStructure) return
-    if (thisReview?.current.reviewStatus === ReviewStatus.Pending) {
+    if (reviewStructure?.thisReview?.current.reviewStatus) {
+      console.log(reviewStructure.thisReview.current.reviewStatus)
+      setReviewStatus(reviewStructure.thisReview.current.reviewStatus)
+    }
```
Comment (CarlosNZ) [inline src/containers/Review/ReviewPage.tsx:81]: If you remove the `useState` variable (previous comment), then you won't need this either.

File: src/containers/Review/ReviewPage.tsx [hunk 3]
```diff
@@ -68,19 +70,30 @@ const ReviewPage: React.FC<{
...
+      console.log('reviewStatus: DISCONTINUED')
       showWarningModal({
-        onConfirm: () => push(`/application/${reviewStructure.info.serial}/review`),
-        showCancel: false,
+        title: strings.REVIEW_STATUS_DISCONTINUED_TITLE,
+        message: strings.REVIEW_STATUS_DISCONTINUED_MESSAGE,
       })
     }
-  }, [reviewStructure])
+  }, [reviewStructure, reviewStatus])
```
Comment (CarlosNZ) [inline src/containers/Review/ReviewPage.tsx:96]: And you won't need it here either, as it gets updated whenever `reviewStructure` changes anyway.

Comment (CarlosNZ) [reply src/containers/Review/ReviewPage.tsx:96]: Fyi -- have tested this without requiring `useState`, and it seems to behave the same.

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Cool, this works nicely. Good job tracking down all (hopefully) the places this is affected by.

Just one main suggestion -- see comments.

File: src/containers/Review/ReviewPage.tsx (hunk 1 above)
Comment (nmadruga) [reply src/containers/Review/ReviewPage.tsx]: Yes, good call. I made this change earlier when I didn't realise there was a problem about the changeStatus action not running correctly. So I haven't change this after to check that. I'll change this back as you suggested. 👍

**Review by CarlosNZ — APPROVED** [review]
Looks good, thanks 👍🏼

**Pull request merged**